### PR TITLE
[expo-av][android] fix for exo player video bandwidth

### PR DESCRIPTION
**Pull request by watchinharrison**
# Why

as described in https://github.com/expo/expo/issues/2177 android videos don't adapt to bandwidth when using an m3u8 adaptive stream file. This means they remain at the lowest quality

# How

Passing through the bandwidthMeter transfer listener to the OkHttpDataSource instance to be set via `addTransferListener`

# Test Plan

View a .m3u8 uri video and see the quality of the video improve, existing videos continue to work as expected

Before:
[Before.mov.zip](https://github.com/expo/expo/files/4652447/Before.mov.zip)

After:
[After.mov.zip](https://github.com/expo/expo/files/4652448/After.mov.zip)

### CODE REVIEW

File: packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/CustomHeadersOkHttpDataSourceFactory.java [hunk 1]
```diff
@@ -24,13 +24,16 @@
...
+  private final TransferListener mBandwidthMeter;
 
-  public CustomHeadersOkHttpDataSourceFactory(@NonNull Call.Factory callFactory, @Nullable String userAgent, @Nullable Map<String, Object> requestHeaders) {
+  public CustomHeadersOkHttpDataSourceFactory(@NonNull Call.Factory callFactory, @Nullable String userAgent, @Nullable Map<String, Object> requestHeaders, @Nullable TransferListener bandwidthMeter) {
     super();
     mCallFactory = callFactory;
     mUserAgent = userAgent;
     mListener = null;
     mCacheControl = null;
+    mBandwidthMeter = bandwidthMeter;
```
Comment (mczernek) [inline packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/CustomHeadersOkHttpDataSourceFactory.java:36]: There already is `mListener` field which is unused, could you please use that one?

**Review by mczernek — COMMENTED** [review]
Thanks a lot for this contribution! That is great!

I will take a closer look at that and either accept it, suggest some changes or, if you don't mind, create own PR based on that, depending on how many changes will I want to involve.

File: packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/DataSourceFactoryProvider.java [hunk 2]
```diff
@@ -3,11 +3,12 @@
...
 import com.google.android.exoplayer2.upstream.DataSource;
+import com.google.android.exoplayer2.upstream.TransferListener;
 
 import java.util.Map;
 
 import org.unimodules.core.ModuleRegistry;
 
 public interface DataSourceFactoryProvider {
-  DataSource.Factory createFactory(Context reactApplicationContext, ModuleRegistry moduleRegistry, String userAgent, Map<String, Object> requestHeaders);
+  DataSource.Factory createFactory(Context reactApplicationContext, ModuleRegistry moduleRegistry, String userAgent, Map<String, Object> requestHeaders, TransferListener bandwidthMeter);
```
Comment (mczernek) [inline packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/DataSourceFactoryProvider.java:13]: I can find two implementations of this interface and only one is changes within this PR. I am afraid Expo Client won't be able to build :(

File: packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/CustomHeadersOkHttpDataSourceFactory.java (hunk 1 above)
Comment (watchinharrison) [reply packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/CustomHeadersOkHttpDataSourceFactory.java]: done 👍

File: packages/expo-av/CHANGELOG.md [hunk 3]
```diff
@@ -9,3 +9,5 @@
 ### 🐛 Bug fixes
 
 - Fixed `Plaback.loadAsync()` return type. ([#7559](https://github.com/expo/expo/pull/7559) by [@awinograd](https://github.com/awinograd))
+
+- Fixed the adaptive streaming for exoplayer for android as raised in https://github.com/expo/expo/issues/2177
```
[the remark this answers is not in the document]
Comment (watchinharrison) [reply packages/expo-av/CHANGELOG.md]: I think my change matches the expected style?

File: packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/DataSourceFactoryProvider.java (hunk 2 above)
Comment (watchinharrison) [reply packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/DataSourceFactoryProvider.java:13]: I think I've found the other impl' and added the expected arg

File: packages/expo-av/CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mczernek) [reply packages/expo-av/CHANGELOG.md]: Now it does, thanks a lot!

File: packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/DataSourceFactoryProvider.java (hunk 2 above)
Comment (mczernek) [reply packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/DataSourceFactoryProvider.java]: There is another one in `versioned.host.exp.exponent.modules.universal.av` package in `:expoview` module.

Comment (watchinharrison) [reply packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/DataSourceFactoryProvider.java]: https://github.com/expo/expo/pull/8363/files#diff-72aabec1f0463289beaa60675f0e93c3 the one here?

Comment (watchinharrison) [reply packages/expo-av/android/src/main/java/expo/modules/av/player/datasource/DataSourceFactoryProvider.java]: sorry found it

**Comment by watchinharrison** [comment]
Changes are included in https://github.com/expo/expo/pull/8380, so closing this PR

**Pull request closed without merging**